keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -381,6 +387,9 @@ but allows infinite retries for pod failures caused by cluster-management
 events (such as preemption). I do not have knowledge or influence over the executable that researchers run,
 so I don't know beforehand which exit codes they might return.
```

Comment:
"infinite retries" - are infinite retries are really desirable even for cluster-management disruptions? It could be due to broken configurations and it might be costly to keep retrying forever. Should we give up users an option to give up even for cluster-management disruptions?

For comparison the current default backoffLimit is 6, so erring on the side of giving up after some time.

Maybe actions could come with retry limits:
action=Retry(x) - equivalent to backoffLimit=x but counted within this rule
action=FailJob - equivalent to Retry(0)
action=Ignore - equivalent to Retry(inf)
traditional backoffLimit in Job spec - equivalent to default Retry(backoffLimit) rule at the end
?
(not sure what to do if two rules want to share the limit, maybe that's not needed)

Reply:
Right, but backoffLimit is also default after not matching any rule. So either way, we won't be able to do "If cluster-management - retry 6 times, everything else - fail immediately [or retry X<6 times]". Have we considered the "else" action? So I could do something like:

backoffLimit: 6
podFailurePolicy:
  rules:
    - action: Count
      onPodConditions:
      - type: DisruptionTarget
   - action: FailJob (else action, by default Count if not provided)

And then the difference between Retry(X) and Count would be just whether we want to support more than one limit.

Reply:
You can do that with `FailJob` on `exit code != 0` :)

Reply:
what about other onPodConditions types? and other future requirements?

Reply:
This:

```
backoffLimit: 6
podFailurePolicy:
  rules:
    - action: Count
      onPodConditions:
      - type: DisruptionTarget
   - action: FailJob
      onExitCodes:
        operator: NotIn
        values: [0]
```

does seem to cover my usecase of avoiding infinite retries: "If transient cluster disruptions - retry X times, everything else - fail immediately". (Consider adding a user story/example to KEP.)

Not sure if more is needed in the API then. If it's easy to add catch-all/else action, I'd still do it. This is natural in "case/when" syntaxes and simpler than figuring out FailJob exit code != 0 trick.

Agree that multiple limits would complicate the API and better to wait for feedback before adding it.

Reply:
I like the Retry(x) idea. I see it as a parametrized Count, where we could indicate which limit we count a Pod failure towards. For example: 

```yaml
podFailurePolicy:
  limits:
    disruptionsLimit: 10
    specialRandomFailureLimit: 4
  rules:
    - action: Count
      countTowardsLimit: disruptionsLimit
      onPodConditions:
      - type: DisruptionTarget
   - action: Count
      countTowardsLimit: specialRandomFailureLimit
      onExitCodes:
        operator: In
        values: [42]
```

However, it complicates the API and I would like to refrain from implementing it until there is an important use case.

For now, I've updated the KEP with Story 3 where I describe the use case "count disruptions but fail jobs immediately on user errors", using the `exit code != 0` trick. 

To avoid the trick I see three solutions (2. was suggested by @qelo internally):
1. keep the API as is, we can provide the configuration in docs so that it will be easy to find out by users
2. allow for the last rule to have an action without `onExitCodes` and `onPodConditions` (it would play a role of `default` in `switch-case` syntax)
3. introduce `fallbackAction` / `defaultAction` just under `podFailurePolicy`

They are in the order of my preference :)

Reply:
We should go with 1 and only 2 if there are requests from the community.

The good thing is that the effect requested by @qelo is fully covered.